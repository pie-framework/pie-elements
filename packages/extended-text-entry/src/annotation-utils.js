// returns DOM Nodes before the stopOffset
const getAllTextNodesBeforePosition = (node, stopOffset, nodeArray) => {
  const nodes = (nodeArray) ? nodeArray : [];
  const offset = nodes.reduce((acc, node) => acc + node.textContent.length, 0);
  let keepWalking = true;

  if (offset > stopOffset) {
    return false;
  }

  // 3 = HTML DOM node value for text nodes
  if (node.nodeType === 3 && node.parentNode && !node.parentNode.hasAttribute('data-ann-id')) {
    nodes.push(node);
  }

  node = node.firstChild;

  while(node && keepWalking) {
    keepWalking = getAllTextNodesBeforePosition(node, stopOffset, nodes);
    node = node.nextSibling;
  }

  return nodes;
};

// returns DOM Nodes that overlap the start or end position
export const getDOMNodes = (startOffset, endOffset, rootNode) => {
  const textNodesToEnd = getAllTextNodesBeforePosition(rootNode, endOffset);
  const domNodes = [];
  let start = 0;

  textNodesToEnd.forEach((node) => {
    const end = start + node.textContent.length;

    [startOffset, endOffset].forEach((offset) => {
      if (start <= offset && offset < end) {
        domNodes.push({
          node,
          offset: offset - start
        });
      }
    });

    start = end;
  });

  return domNodes;
};

// returns [ textNodes, text ]
// textNodes = DOM Text Nodes between start range and end range node
// text = plain text in range
const getTextNodesBetween = (range) => {
  const {
    commonAncestorContainer: rootNode,
    startContainer: startNode,
    endContainer: endNode,
    startOffset,
    endOffset
  } = range;
  let pastStartNode = false;
  let reachedEndNode = false;
  const textNodes = [];
  let text = '';

  const getTextNodes = (node) => {
    const { nodeValue: value, childNodes } = node;

    if(node === startNode && node === endNode) {
      if (value) {
        text += value.substring(startOffset, endOffset);
      }
      pastStartNode = reachedEndNode = true;
    } else if(node === startNode) {
      if (value) {
        text += value.substring(startOffset);
      }
      pastStartNode = true;
    } else if(node === endNode) {
      if (value) {
        text += value.substring(0, endOffset);
      }
      reachedEndNode = true;
    } else if(node.nodeType === 3 && node.parentNode && !node.parentNode.hasAttribute('data-ann-id')) {
      // 3 = HTML DOM node value for text nodes
      if (value && pastStartNode && !reachedEndNode) {
        text += value;
        textNodes.push(node);
      }
    }

    childNodes.forEach((childNode) => {
      if (!reachedEndNode) {
        getTextNodes(childNode);
      } else {
        return;
      }
    });
  };

  getTextNodes(rootNode);
  return [textNodes, text];
};

const surroundContent = range => {
  const wrapper = document.createElement('SPAN');

  range.surroundContents(wrapper);

  return wrapper;
};

// wrap each DOM Text Node in range into a span
export const wrapRange = (range) => {
  if (range.startContainer === range.endContainer) {
    return [ surroundContent(range) ];
  }

  const [ nodesBetween ] = getTextNodesBetween(range);

  // wrap the start node
  const startRange = document.createRange();
  startRange.selectNodeContents(range.startContainer);
  startRange.setStart(range.startContainer, range.startOffset);
  const startWrapper = surroundContent(startRange);

  // wrap the end node
  const endRange = document.createRange();
  endRange.selectNode(range.endContainer);
  endRange.setEnd(range.endContainer, range.endOffset);
  const endWrapper = surroundContent(endRange);

  // wrap the nodes between start and end nodes, if any
  const centerWrappers = nodesBetween.map((node) => {
    const wrapper = document.createElement('SPAN');

    node.parentNode.insertBefore(wrapper, node);
    wrapper.appendChild(node);

    return wrapper;
  });

  return [ startWrapper, ...centerWrappers, endWrapper ];
};

// returns text in range with start and end position in rootNode
export const getRangeDetails = (range, rootNode) => {
  const rangeBefore = document.createRange();
  rangeBefore.setStart(rootNode, 0);
  rangeBefore.setEnd(range.startContainer, range.startOffset);

  const [ , text ] = getTextNodesBetween(rangeBefore);
  const [ , textSelected ] = getTextNodesBetween(range);
  const start = text.length;
  const end = start + textSelected.length;

  return {
    quote: textSelected,
    start,
    end
  };
};

// removes the annotation spans without changing the content inside
export const removeElemsWrapping = (elems, container) => {
  elems.forEach(elem => {
    const parent = elem.parentNode;
    const childNodes = elem.childNodes;
    const childNodesLength = childNodes.length;

    if (childNodesLength > 0) {
      for (let i = 0; i < childNodesLength; i++) {
        parent.insertBefore(childNodes[0], elem);
      }
    } else {
      parent.insertBefore(document.createTextNode(elem.textContent), elem);
    }

    parent.removeChild(elem);
  });

  container.normalize();
};

// also used in select-text
export const clearSelection = () => {
  if (document.getSelection) {
    // for all new browsers (IE9+, Chrome, Firefox)
    document.getSelection().removeAllRanges();
    document.getSelection().addRange(document.createRange());
  } else if (window.getSelection) {
    // equals with the document.getSelection (MSDN info)
    if (window.getSelection().removeAllRanges) {
      // for all new browsers (IE9+, Chrome, Firefox)
      window.getSelection().removeAllRanges();
      window.getSelection().addRange(document.createRange());
    } else if (window.getSelection().empty) {
      // Chrome supports this as well
      window.getSelection().empty();
    }
  } else if (document.selection) {
    // IE8-
    document.selection.empty();
  }
};

// check if annotation label contains line breaks or its length >= 20 characters
export const isSideLabel = text => text.length >= 20 || text.search(/\n|\r|\r\n/) !== -1;

export const getAnnotationElements = id => {
  return Array.from(document.querySelectorAll(`[data-id='${id}']`));
};

export const getLabelElement = id => {
  return document.querySelector(`[data-ann-id='${id}']`);
};
