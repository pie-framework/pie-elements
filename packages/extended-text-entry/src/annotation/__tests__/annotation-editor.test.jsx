import { shallow } from 'enzyme';
import React from 'react';
import AnnotationEditor from '../annotation-editor';
import { unwrap } from "@material-ui/core/test-utils";
import * as utils from '../annotation-utils';

const Editor = unwrap(AnnotationEditor);
const predefinedAnnotations = [
  {
    label: 'good',
    text: 'good',
    type: 'positive'
  }, {
    label: '★',
    text: '★',
    type: 'positive'
  }, {
    label: 'cut',
    text: 'cut',
    type: 'negative'
  }, {
    label: 'sp',
    text: 'spelling',
    type: 'negative'
  }
];
const mockedElementDOM = {
  removeChild: jest.fn(),
  classList: {
    remove: jest.fn(),
    add: jest.fn()
  },
};

jest.spyOn(utils, 'removeElemsWrapping').mockImplementation(() => {});

describe('freeform editor', () => {
  const defaultProps = {
    classes: {},
    text: '',
    annotations: [],
    comment: '',
    predefinedAnnotations: [],
    width: 500,
    height: 100,
    maxHeight: '40vh',
    disabled: false,
    disabledMath: false,
    customKeys: [],
    keypadMode: 'miscellaneous',
  };
  let onChange;
  let onCommentChange;

  beforeEach(() => {
    onChange = jest.fn();
    onCommentChange = jest.fn();

    global.document.createRange = () => ({
      setStart: () => {},
      setEnd: () => {},
      commonAncestorContainer: {
        nodeName: 'BODY',
        ownerDocument: document,
      },
    });
  });

  const wrapper = extras => {
    const defaults = {
      ...defaultProps,
      onChange,
      onCommentChange
    };
    const props = { ...defaults, ...extras };

    return shallow(<Editor { ...props } />);
  };

  describe('snapshots', () => {
    it('renders', () => {
      expect(wrapper()).toMatchSnapshot();
    });

    it('renders disabled', () => {
      expect(wrapper({ disabled: true })).toMatchSnapshot();
    });

    it('renders with predefined annotations', () => {
      expect(wrapper({ predefinedAnnotations })).toMatchSnapshot();
    });

    it('renders with text and annotations', () => {
      expect(
        wrapper({
          text: '<div>This is an example.</div>',
          annotations: [
            {
              'id': '1',
              'label': 'good',
              'type': 'positive',
              'quote': 'This',
              'start': 0,
              'end': 4
            }
          ]
        })
      ).toMatchSnapshot();
    });

    it('renders with complex text and nested annotations', () => {
      expect(
        wrapper({
          text: '<div>This is <b>a complex </b><em>example</em><br/>This is a new line.</div>',
          annotations: [
            {
              'id': '1',
              'label': 'creative',
              'type': 'positive',
              'quote': 'This is a complex exam',
              'start': 0,
              'end': 22
            },
            {
              'id': '2',
              'label': 'punctuation',
              'type': 'negative',
              'quote': 'exampleThis is',
              'start': 18,
              'end': 32
            },
            {
              'id': '3',
              'label': 'cut',
              'type': 'negative',
              'quote': 'a comp',
              'start': 8,
              'end': 14
            }
          ]
        })
      ).toMatchSnapshot();
    });

    it('renders with text and comment', () => {
      expect(
        wrapper({
          text: '<div>This is an example.</div>',
          comment: '<div>Very good</div>'
        })
      ).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('select a predefined annotation', () => {
      it('calls onChange when new annotation selected',  () => {
        const w = wrapper({
          text: '<div>This is an example.</div>',
        });

        w.instance().createDOMAnnotation = jest.fn();
        w.instance().handleClose = jest.fn();
        w.update();

        w.setState({
          selectedElems: [mockedElementDOM, mockedElementDOM],
          selectionDetails: {
            'quote': 'This',
            'start': 0,
            'end': 4
          },
        });
        w.instance().handleMenuClick({
          type: 'positive',
          text: 'good'
        });

        expect(w.instance().createDOMAnnotation).toBeCalled();
        expect(onChange).toBeCalled();
        expect(w.instance().handleClose).toBeCalled();
      });

      it('calls onChange when old annotation selected',  () => {
        const w = wrapper({
          text: '<div>This is an example.</div>',
          annotations: [
            {
              'id': '1',
              'label': 'good',
              'type': 'positive',
              'quote': 'This',
              'start': 0,
              'end': 4
            }
          ]
        });

        w.instance().updateLabel = jest.fn();
        w.instance().handleClose = jest.fn();
        w.update();

        w.setState({
          annotation: {
            'id': '1',
            'label': 'good',
            'type': 'positive',
            'quote': 'This',
            'start': 0,
            'end': 4
          },
          labelElem: mockedElementDOM,
          selectedElems: [mockedElementDOM, mockedElementDOM],
          annotationIndex: 0
        });
        w.instance().handleMenuClick({
          type: 'negative',
          text: 'wrong'
        });
        expect(w.instance().updateLabel).toBeCalled();
        expect(onChange).toBeCalled();
        expect(w.instance().handleClose).toBeCalled();
      });
    });

    describe('write new annotation', () => {
      it('calls onChange',  () => {
        const w = wrapper({
          text: '<div>This is an example.</div>',
        });

        w.instance().createDOMAnnotation = jest.fn();
        w.update();

        w.setState({
          selectedElems: [mockedElementDOM, mockedElementDOM],
          selectionDetails: {
            'quote': 'This',
            'start': 0,
            'end': 4
          },
        });
        w.instance().addAnnotation('positive');

        expect(w.instance().createDOMAnnotation).toBeCalled();
        expect(onChange).toBeCalled();
      });
    });

    describe('edit annotation', () => {
      it('updates state',  () => {
        const w = wrapper();

        w.setState({
          openedMenu: true,
          openedEditor: false
        });
        w.instance().editAnnotation();
        expect(w.state().openedMenu).toEqual(false);
        expect(w.state().openedEditor).toEqual(true);
      });
    });

    describe('delete annotation', () => {
      it('calls onChange',  () => {
        const w = wrapper({
          text: '<div>This is an example.</div>',
        });

        w.instance().handleClose = jest.fn();
        w.update();

        w.setState({
          annotation: {
            'id': '1',
            'label': 'good',
            'type': 'positive',
            'quote': 'This',
            'start': 0,
            'end': 4
          },
          labelElem: mockedElementDOM,
          selectedElems: [mockedElementDOM, mockedElementDOM],
          annotationIndex: 0
        });
        w.instance().deleteAnnotation();

        expect(utils.removeElemsWrapping).toBeCalled();
        expect(onChange).toBeCalled();
        expect(w.instance().handleClose).toBeCalled();
      });
    });

    describe('update annotation', () => {
      it('calls onChange',  () => {
        const w = wrapper({
          text: '<div>This is an example.</div>',
          annotations: [
            {
              'id': '1',
              'label': 'good',
              'type': 'positive',
              'quote': 'This',
              'start': 0,
              'end': 4
            }
          ]
        });

        w.instance().updateLabel = jest.fn();
        w.update();

        w.setState({
          annotationIndex: 0
        });
        w.instance().updateAnnotation();

        expect(w.instance().updateLabel).toBeCalled();
        expect(onChange).toBeCalled();
      });
    });

    describe('change annotation type', () => {
      it('calls onChange',  () => {
        const w = wrapper({
          text: '<div>This is an example.</div>',
          annotations: [
            {
              'id': '1',
              'label': 'good',
              'type': 'positive',
              'quote': 'This',
              'start': 0,
              'end': 4
            }
          ]
        });

        w.instance().updateLabel = jest.fn();
        w.instance().handleClose = jest.fn();
        w.update();

        w.setState({
          annotation: {
            'id': '1',
            'label': 'good',
            'type': 'positive',
            'quote': 'This',
            'start': 0,
            'end': 4
          },
          labelElem: mockedElementDOM,
          selectedElems: [mockedElementDOM, mockedElementDOM],
          annotationIndex: 0
        });
        w.instance().changeAnnotationType('very good');
        expect(w.instance().updateLabel).toBeCalled();
        expect(onChange).toBeCalled();
        expect(w.instance().handleClose).toBeCalled();
      });
    });

    describe('update annotation label', () => {
      it('calls createDOMAnnotation',  () => {
        const w = wrapper({
          text: '<div>This is an example.</div>',
          annotations: [
            {
              'id': '1',
              'label': 'good',
              'type': 'positive',
              'quote': 'This',
              'start': 0,
              'end': 4
            }
          ]
        });

        w.instance().updateLabel = jest.fn();
        w.instance().handleClose = jest.fn();
        w.update();

        w.setState({
          annotation: {
            'id': '1',
            'label': 'good',
            'type': 'positive',
            'quote': 'This',
            'start': 0,
            'end': 4
          },
          labelElem: mockedElementDOM,
          selectedElems: [mockedElementDOM, mockedElementDOM],
          annotationIndex: 0
        });
        w.instance().changeAnnotationType('very good');
        expect(w.instance().updateLabel).toBeCalled();
        expect(onChange).toBeCalled();
        expect(w.instance().handleClose).toBeCalled();
      });
    });

  });
});
