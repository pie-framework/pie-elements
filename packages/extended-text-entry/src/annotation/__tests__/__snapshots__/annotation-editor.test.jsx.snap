// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`freeform editor snapshots renders 1`] = `
<div>
  <div>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "maxHeight": "40vh",
          "minHeight": 100,
          "width": 466,
        }
      }
    />
    <div />
  </div>
  <WithStyles(RawInputContainer)
    label="Comment"
  >
    <EditableHtml
      allowValidation={false}
      disabled={false}
      markup=""
      onChange={[MockFunction]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "math": Object {
            "controlledKeypadMode": false,
            "customKeys": Array [],
            "disabled": false,
            "keypadMode": "miscellaneous",
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
      width="604"
    />
  </WithStyles(RawInputContainer)>
  <WithStyles(AnnotationMenu)
    anchorEl={null}
    annotations={Array []}
    isNewAnnotation={false}
    onAnnotate={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onWrite={[Function]}
    open={false}
  />
  <WithStyles(FreeformEditor)
    offset={0}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    onTypeChange={[Function]}
    open={false}
    type={null}
    value=""
  />
</div>
`;

exports[`freeform editor snapshots renders disabled 1`] = `
<div>
  <div>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "maxHeight": "40vh",
          "minHeight": 100,
          "width": 466,
        }
      }
    />
    <div />
  </div>
  <WithStyles(RawInputContainer)
    label="Comment"
  >
    <EditableHtml
      allowValidation={false}
      disabled={true}
      markup=""
      onChange={[MockFunction]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "math": Object {
            "controlledKeypadMode": false,
            "customKeys": Array [],
            "disabled": false,
            "keypadMode": "miscellaneous",
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
      width="604"
    />
  </WithStyles(RawInputContainer)>
  <WithStyles(AnnotationMenu)
    anchorEl={null}
    annotations={Array []}
    isNewAnnotation={false}
    onAnnotate={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onWrite={[Function]}
    open={false}
  />
  <WithStyles(FreeformEditor)
    offset={0}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    onTypeChange={[Function]}
    open={false}
    type={null}
    value=""
  />
</div>
`;

exports[`freeform editor snapshots renders with complex text and nested annotations 1`] = `
<div>
  <div>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<div>This is <b>a complex </b><em>example</em><br/>This is a new line.</div>",
        }
      }
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "maxHeight": "40vh",
          "minHeight": 100,
          "width": 466,
        }
      }
    />
    <div />
  </div>
  <WithStyles(RawInputContainer)
    label="Comment"
  >
    <EditableHtml
      allowValidation={false}
      disabled={false}
      markup=""
      onChange={[MockFunction]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "math": Object {
            "controlledKeypadMode": false,
            "customKeys": Array [],
            "disabled": false,
            "keypadMode": "miscellaneous",
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
      width="604"
    />
  </WithStyles(RawInputContainer)>
  <WithStyles(AnnotationMenu)
    anchorEl={null}
    annotations={Array []}
    isNewAnnotation={false}
    onAnnotate={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onWrite={[Function]}
    open={false}
  />
  <WithStyles(FreeformEditor)
    offset={0}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    onTypeChange={[Function]}
    open={false}
    type={null}
    value=""
  />
</div>
`;

exports[`freeform editor snapshots renders with predefined annotations 1`] = `
<div>
  <div>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "maxHeight": "40vh",
          "minHeight": 100,
          "width": 466,
        }
      }
    />
    <div />
  </div>
  <WithStyles(RawInputContainer)
    label="Comment"
  >
    <EditableHtml
      allowValidation={false}
      disabled={false}
      markup=""
      onChange={[MockFunction]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "math": Object {
            "controlledKeypadMode": false,
            "customKeys": Array [],
            "disabled": false,
            "keypadMode": "miscellaneous",
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
      width="604"
    />
  </WithStyles(RawInputContainer)>
  <WithStyles(AnnotationMenu)
    anchorEl={null}
    annotations={
      Array [
        Object {
          "label": "good",
          "text": "good",
          "type": "positive",
        },
        Object {
          "label": "★",
          "text": "★",
          "type": "positive",
        },
        Object {
          "label": "cut",
          "text": "cut",
          "type": "negative",
        },
        Object {
          "label": "sp",
          "text": "spelling",
          "type": "negative",
        },
      ]
    }
    isNewAnnotation={false}
    onAnnotate={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onWrite={[Function]}
    open={false}
  />
  <WithStyles(FreeformEditor)
    offset={0}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    onTypeChange={[Function]}
    open={false}
    type={null}
    value=""
  />
</div>
`;

exports[`freeform editor snapshots renders with text and annotations 1`] = `
<div>
  <div>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<div>This is an example.</div>",
        }
      }
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "maxHeight": "40vh",
          "minHeight": 100,
          "width": 466,
        }
      }
    />
    <div />
  </div>
  <WithStyles(RawInputContainer)
    label="Comment"
  >
    <EditableHtml
      allowValidation={false}
      disabled={false}
      markup=""
      onChange={[MockFunction]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "math": Object {
            "controlledKeypadMode": false,
            "customKeys": Array [],
            "disabled": false,
            "keypadMode": "miscellaneous",
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
      width="604"
    />
  </WithStyles(RawInputContainer)>
  <WithStyles(AnnotationMenu)
    anchorEl={null}
    annotations={Array []}
    isNewAnnotation={false}
    onAnnotate={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onWrite={[Function]}
    open={false}
  />
  <WithStyles(FreeformEditor)
    offset={0}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    onTypeChange={[Function]}
    open={false}
    type={null}
    value=""
  />
</div>
`;

exports[`freeform editor snapshots renders with text and comment 1`] = `
<div>
  <div>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<div>This is an example.</div>",
        }
      }
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "maxHeight": "40vh",
          "minHeight": 100,
          "width": 466,
        }
      }
    />
    <div />
  </div>
  <WithStyles(RawInputContainer)
    label="Comment"
  >
    <EditableHtml
      allowValidation={false}
      disabled={false}
      markup="<div>Very good</div>"
      onChange={[MockFunction]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "math": Object {
            "controlledKeypadMode": false,
            "customKeys": Array [],
            "disabled": false,
            "keypadMode": "miscellaneous",
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
      width="604"
    />
  </WithStyles(RawInputContainer)>
  <WithStyles(AnnotationMenu)
    anchorEl={null}
    annotations={Array []}
    isNewAnnotation={false}
    onAnnotate={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onWrite={[Function]}
    open={false}
  />
  <WithStyles(FreeformEditor)
    offset={0}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    onTypeChange={[Function]}
    open={false}
    type={null}
    value=""
  />
</div>
`;
